clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -133,6 +144,25 @@ public Struct toStruct() {
     }
 
     @Override
+    public String toString() {
+        // Store it in a local variable to protect against concurrent updates
+        Map<TopicPartition, MemoryRecords> partitionRecords = this.partitionRecords;
+
+        StringBuilder bld = new StringBuilder();
+        bld.append("(type=ProduceRequest")
+                .append(", acks=").append(acks)
+                .append(", timeout=").append(timeout);
+
+        if (partitionRecords == null)
+            bld.append(", partitions=").append(Utils.mkString(partitions, ","));
+        else
+            bld.append(", partitionRecords=(").append(Utils.mkString(partitionRecords));
```

Comment:
Thanks for the feedback. I'll try to address the comments:

1. Formatting differs from Struct's. That's right, I used the same format as ProduceRequest.Builder as that was easier, but I can use Struct's formatting if people prefer that (there's some inconsistency, either way).

2. It prints a summary of the record contents (i.e. `Record(magic = %d, attributes = %d, compression = %s, crc = %d, %s = %d, key = %d bytes, value = %d bytes)` since that's the current behaviour in trunk. Note that the old Scala class only printed the partitions and sizes if TRACE logging was enabled. For DEBUG logging, the partitions information wasn't printed at all. We lost that distinction with the Java classes. I can fix that for Fetch and Produce requests in this PR and we can tackle the rest later. It seems to me that printing the record summary if TRACE logging is enabled would make sense. However, since the information won't be available in many/most cases, it makes sense to keep it simple and never print more than the size.